src/managers/QueueManager.sol

```diff
@@ -0,0 +1,113 @@
+///      (e.g. if the on/off ramp manager is used, or if sync deposits are enabled). This prevents spam.
+contract QueueManager is IQueueManager, IUpdateContract {
+    using CastLib for *;
+    using BitmapLib for *;
+
+    address public immutable contractUpdater;
+    IBalanceSheet public immutable balanceSheet;
+
+    mapping(PoolId => mapping(ShareClassId => ShareClassQueueState)) public scQueueState;
+
+    constructor(address contractUpdater_, IBalanceSheet balanceSheet_) {
+        contractUpdater = contractUpdater_;
+        balanceSheet = balanceSheet_;
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Owner actions
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IUpdateContract
+    function update(PoolId poolId, ShareClassId scId, bytes calldata payload) external {
+        require(msg.sender == contractUpdater, NotContractUpdater());
+
+        uint8 kind = uint8(UpdateContractMessageLib.updateContractType(payload));
+        if (kind == uint8(UpdateContractType.UpdateQueue)) {
+            UpdateContractMessageLib.UpdateContractUpdateQueue memory m =
+                UpdateContractMessageLib.deserializeUpdateContractUpdateQueue(payload);
+            ShareClassQueueState storage sc = scQueueState[poolId][scId];
+            sc.minDelay = m.minDelay;
+            sc.extraGasLimit = m.extraGasLimit;
+            emit UpdateQueueConfig(poolId, scId, m.minDelay, m.extraGasLimit);
+        } else {
+            revert UnknownUpdateContractType();
+        }
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Sync
+    //----------------------------------------------------------------------------------------------
+
+    /// @inheritdoc IQueueManager
+    function sync(PoolId poolId, ShareClassId scId, AssetId[] calldata assetIds) external {
+        ShareClassQueueState storage sc = scQueueState[poolId][scId];
+        require(
+            sc.lastSync == 0 || sc.minDelay == 0 || block.timestamp >= sc.lastSync + sc.minDelay, MinDelayNotElapsed()
+        );
+
+        (uint128 delta,, uint32 queuedAssetCounter,) = balanceSheet.queuedShares(poolId, scId);
+        require(delta > 0 || queuedAssetCounter > 0, NoUpdates());
+
+        require(assetIds.length <= 256, TooManyAssets()); // Bitmap limit
+
+        // Deduplicate and validate using bitmap for valid indices
+        uint256 validBitmap = 0; // Each bit represents if that index is valid
+        uint256 validCount = 0;
+
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            bytes32 key = bytes32(uint256(AssetId.unwrap(assetIds[i])));
+            if (TransientStorageLib.tloadBool(key)) {
+                continue; // Skip duplicate
+            }
+            TransientStorageLib.tstore(key, true);
```

Comment:
I think this is no longer need it, can we not check the bitmap `validBitmap`?

Reply:
I realized the Bitmap helps with avoiding the noop, but it still needs to check duplicates. I didn't see a way to check duplicates with the Bitmap alone

Reply:
Yeah I was also missing that a bit. Discussed with Onno and to the best of my knowledge, transient is the most readable gas-efficient duplicate check.